### PR TITLE
chore: Adds workflow that generates changelog after PR is merged

### PR DESCRIPTION
**Pull request by oarbusi**
## Description

Adds a workflow to automatically update the unreleased section of the changelog after a PR with a new changelog entry is merged or there is a new entry in the changelog in the repository and the workflow is triggered manually.
Link to any related issue(s): CLOUDP-241727

## Type of change:

- [ ] Bug fix (non-breaking change which fixes an issue). Please, add the "bug" label to the PR.
- [ ] New feature (non-breaking change which adds functionality). Please, add the "enhancement" label to the PR.
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected). Please, add the "breaking change" label to the PR.
- [ ] This change requires a documentation update
- [ ] Documentation fix/enhancement

## Required Checklist:

- [x] I have signed the [MongoDB CLA](https://www.mongodb.com/legal/contributor-agreement)
- [x] I have read the [contributing guides](https://github.com/mongodb/terraform-provider-mongodbatlas/blob/master/contributing/README.md)
- [x] I have checked that this change does not generate any credentials and that **they are NOT accidentally logged anywhere**.
- [x] I have added tests that prove my fix is effective or that my feature works per HashiCorp requirements
- [x] I have added any necessary documentation (if appropriate)
- [x] I have run make fmt and formatted my code
- [x] If changes include deprecations or removals, I defined an isolated PR with a relevant title as it will be used in the auto-generated changelog.
- [x] If changes include removal or addition of 3rd party GitHub actions, I updated our internal document. Reach out to the APIx Integration slack channel to get access to the internal document.

## Further comments
Tested the workflow in a [fork](https://github.com/oarbusi/terraform-provider-mongodbatlas)
- Manual trigger of workflow: https://github.com/oarbusi/terraform-provider-mongodbatlas/commit/f717afa11bc7505aafb3fdf07e050bc97fb4e0e0
- Automatic trigger of workflow after PR is merged: https://github.com/oarbusi/terraform-provider-mongodbatlas/commit/bc819700f7be137ccf2aa756442c6cf7c6c0a59d
- Full commit history: https://github.com/oarbusi/terraform-provider-mongodbatlas/commits/master/

### CODE REVIEW

File: contributing/changelog-process.md
Comment (oarbusi) [inline contributing/changelog-process.md]: link will not work until this PR is merged

File: .github/workflows/generate-changelog.yml
Comment (lantoli) [inline .github/workflows/generate-changelog.yml:11]: knitpit: this is strictly correct and efficient because it's not run if a PR is closed but not merged.

alternatively if we prefer it to be simpler we can remove that if condition.  It'll run sometimes when it's not necessary (when a pr is closed without merge) but there is not problem as this action is idempotent.

Comment (lantoli) [inline .github/workflows/generate-changelog.yml:11]: knitpit: this is strictly correct and efficient because it's not run if a PR is closed but not merged.

alternatively if we prefer it to be simpler we can remove that if condition.  It'll run sometimes when it's not necessary (when a pr is closed without merge) but there is not problem as this action is idempotent.

File: .github/workflows/generate-changelog.yml
Comment (lantoli) [inline .github/workflows/generate-changelog.yml:2]: @oarbusi cc @AgustinBettati as this is creating a commit in master there might be a race condition, e.g. if 2 PRs are merged almost at the same time.  I know this is highly unlikely but also it's very easy to avoid.
e.g.: 
- PR 1 is merged
- job 1 starts and get current status with only entry file 1
- PR 2 is merged
- job 2 starts, get 1 and 2, and it's faster than job 1 and commits
- job 1 commits, info about PR 2 is lost

We might want to use concurrency limit for all GH actions writing to master: https://docs.github.com/en/actions/using-jobs/using-concurrency

e.g.:

concurrency:
  group: master-write
  cancel-in-progress: false

Comment (lantoli) [reply .github/workflows/generate-changelog.yml:2]: (for example release process might also use this concurrency group)

File: .github/workflows/generate-changelog.yml
Comment (lantoli) [inline .github/workflows/generate-changelog.yml]: we might want to use a MongoDB email address

File: .github/workflows/generate-changelog.yml
Comment (lantoli) [reply .github/workflows/generate-changelog.yml:11]: also similarly to example.yaml we can use something like:

  pull_request:
 ...
    paths:
      - .changelog/**

so it will run only if the PR has an associated changelo entry file

but as my previous comment, I don't think it's very important to optimize run of this GH action

Comment (oarbusi) [reply .github/workflows/generate-changelog.yml:11]: I would avoid removing the if just in case a PR is closed with a new file in .changelog directory which could cause issues in the changelog

File: .github/workflows/generate-changelog.yml
Comment (lantoli) [inline .github/workflows/generate-changelog.yml:32]: what happens if there are not entries yet? (e.g. in a new release and the PR is a chore with no entry file)

is an empty commit pushed?

**Review by lantoli — APPROVED** [review]
LGTM, some minor comments

File: .github/workflows/generate-changelog.yml
Comment (lantoli) [reply .github/workflows/generate-changelog.yml:11]: doesn't it run with latest master content?

File: .github/workflows/generate-changelog.yml
Comment (oarbusi) [reply .github/workflows/generate-changelog.yml:2]: I might be wrong but in the case you explained:

> - job 2 starts, get 1 and 2, and it's faster than job 1 and commits

here changes from PR 1 and 2 are commited (and pushed)

> - job 1 commits, info about PR 2 is lost

job 1 would fail to push because tip of current branch would be behind its remote counterpart

so I think the issue would not happen in this specific case.
If the order was reversed (job 1 commits and pushes before job 2) the push from job 2 would fail and PR2 info would be "lost". This would be easily fixable by manually triggering the workflow.
All in all I think it a very unlikely edge case but we can use concurrency as you suggested since it has an easy fix

File: .github/workflows/generate-changelog.yml
Comment (oarbusi) [reply .github/workflows/generate-changelog.yml]: I used the [same](https://github.com/hashicorp/terraform-provider-aws/blob/60cf7fd65c391de1e8fc200207131eb0f35fe028/.github/workflows/generate_changelog.yml#L26) as AWS uses in their provider. Happy to change it. 
Would `changelogbot@mongodb.com` be better?

File: .github/workflows/generate-changelog.yml
Comment (AgustinBettati) [inline .github/workflows/generate-changelog.yml]:
```suggestion
  generate-changelog:
```

File: .github/workflows/generate-changelog.yml
Comment (oarbusi) [reply .github/workflows/generate-changelog.yml:32]: Nothing would happen. 
`update-changelog-unreleased-section.sh` only updates changelog if there are new entries. If the script does not do any changes to changelog the script would not enter inside the outermost if where the commit and push are performed`if [[ $(git status --porcelain) ]]; then`

File: contributing/changelog-process.md
Comment (AgustinBettati) [inline contributing/changelog-process.md]:
```suggestion
After a PR is merged to master with a new entry in `.changelog` directory, [Generate CHANGELOG workflow](https://github.com/mongodb/terraform-provider-mongodbatlas/actions/workflows/generate-changelog.yml) will be triggered and it will update the `CHANGELOG.md` file with the new entry.
```

File: contributing/changelog-process.md
Comment (AgustinBettati) [inline contributing/changelog-process.md:156]: We can add this title in the index at the top of the file so it has visibility.

File: .github/workflows/generate-changelog.yml
Comment (oarbusi) [reply .github/workflows/generate-changelog.yml:11]: yes you are correct, my bad. Even if it's run in latest master branch, I think it's better to keep the if. It will avoid  unnecessary jobs being triggered and help to keep the action history cleaner

File: contributing/changelog-process.md
Comment (oarbusi) [reply contributing/changelog-process.md]: This made me realise that with current implementation, the action would be triggered if a PR is merged into any branch (not only master). Thanks! [Fixed here](https://github.com/mongodb/terraform-provider-mongodbatlas/pull/2128/commits/25e59b9483d71887ca1fcde827f4828671d10af0)

File: .github/workflows/generate-changelog.yml
Comment (lantoli) [reply .github/workflows/generate-changelog.yml:11]: ok, then you might also want to add "paths" so it's only run when there is a changelog file as most of PRs won't have a changelog file

File: contributing/changelog-process.md
Comment (oarbusi) [reply contributing/changelog-process.md:156]: Done!

File: .github/workflows/generate-changelog.yml
Comment (lantoli) [reply .github/workflows/generate-changelog.yml:2]: you're right push would fail.

you brought an interesting point, how can we know if this GH action fails?  if it runs and fail i don't think we'll be watching (manually) if they're working

File: .github/workflows/generate-changelog.yml
Comment (lantoli) [reply .github/workflows/generate-changelog.yml]: not sure what email it's best to use, maybe @gssbzn can help.
as an example I found in CLI:
maintainer: MongoDB Inc. <support@mongodb.com>

File: contributing/changelog-process.md
Comment (lantoli) [reply contributing/changelog-process.md]: what about doing something similar to examples.yaml and run on push instead of pull_request:

push:
    branches:
      - master

File: .github/workflows/generate-changelog.yml
Comment (lantoli) [inline .github/workflows/generate-changelog.yml]:
```suggestion
name: Generate Changelog
```

File: contributing/changelog-process.md
Comment (marcosuma) [inline contributing/changelog-process.md]:
```suggestion
This workflow can also be triggered manually and it will update the `CHANGELOG.md` file with all entries in `.changelog` directory that are not present in the `CHANGELOG.md` file.
```

File: .github/workflows/generate-changelog.yml
Comment (oarbusi) [reply .github/workflows/generate-changelog.yml:11]: Done!

**Review by EspenAlbert — APPROVED** [review]

File: .github/workflows/generate-changelog.yml
Comment (AgustinBettati) [reply .github/workflows/generate-changelog.yml:2]: To make sure I understand, the concurrency group configuration would prevent this potential push failure as one wound run after the other right?
It is true we will likely not keep an eye on the execution of this workflow, would a slack message to our notifications channel be sufficient? Not suggesting to do as part of this PR of course.

Comment (oarbusi) [reply .github/workflows/generate-changelog.yml:2]: Good point, we won't be actively looking at this GH action failing.
I see two ways of avoiding forgetting about this GH action failing
- Add a step to notify team/oncall person when it fails
- Trigger this workflow as part of the release process/workflow. 

I feel the second option is better because we avoid disrupting team/oncall with a notification where the fix is trivial(retrigger workflow) and we automate the fix. If in the release process the step fails we would notice it because when releasing whoever is working on it will notice and will fix it.
WDYT?
cc. @AgustinBettati

Comment (AgustinBettati) [reply .github/workflows/generate-changelog.yml:2]: I like the idea of keeping the changelog up-to-date after each PR as it will update the `Unreleased` section of changelog (helpful for us and external users).
I would not anticipate this workflow having frequent failures, meaning it would no be disruptive for on call. Seeing [cloudflares history](https://github.com/cloudflare/terraform-provider-cloudflare/actions/workflows/generate-changelog.yml) for this action the only failures I see is because of `Updates were rejected because the remote contains work that you do not have locally` (exact issue we where discussing here).

File: .github/workflows/generate-changelog.yml
Comment (gssbzn) [reply .github/workflows/generate-changelog.yml]: We have cloud bot (https://github.com/mms-build-account) but you have to consider in the near future all shipped product needs it's commit signed according to SSDLC cc @marcosuma

File: .github/workflows/generate-changelog.yml
Comment (lantoli) [reply .github/workflows/generate-changelog.yml:2]: @AgustinBettati yes, concurrency is like a mutex so only one job is executed and the others wait (similar to the mutex in ProjectIDExecution).   

i don't dislike the idea of notifying in Slack when this fails as i don't anticipate many errors either, and it can help us to understand better the problems.

and it might be useful for other GH actions that are also unattended

Comment (lantoli) [reply .github/workflows/generate-changelog.yml:2]: about "Trigger this workflow as part of the release process/workflow" we could do that in addition to do it in each PR in case the latest ones failed, but it's probably not worth it and it's better just to get notified if it fails

File: .github/workflows/generate-changelog.yml
Comment (lantoli) [reply .github/workflows/generate-changelog.yml]: (wow, i'm surprised AWS uses that address)

File: .github/workflows/generate-changelog.yml
Comment (oarbusi) [reply .github/workflows/generate-changelog.yml:2]: Thanks for the input! Seems like we reached a decision: Notifying through slack when the job fails. I can do it in a follow up PR. Thanks again

File: .github/workflows/generate-changelog.yml
Comment (oarbusi) [reply .github/workflows/generate-changelog.yml]: @gssbzn what does _near future_ mean in this context? so we have an idea if we have to address this now or a bit later.
 @AgustinBettati for context

Comment (gssbzn) [reply .github/workflows/generate-changelog.yml]: June next year all ship products need by be compliant with SSDLC

File: .github/workflows/generate-changelog.yml
Comment (lantoli) [reply .github/workflows/generate-changelog.yml:2]: ok not to implement concurrency at the moment and see if we get errors

File: .github/workflows/generate-changelog.yml
Comment (AgustinBettati) [reply .github/workflows/generate-changelog.yml]: We can make use of ghaction-import-gpg action to sign commits https://github.com/actions/runner/issues/667#issuecomment-775277257. We currently use this action for signing our release. @oarbusi We can include in a followup PR but seems reasonable.

Regarding
> Would changelogbot@mongodb.com be better?

I agree.

**Pull request merged**
